src/main/java/com/imaee/propinq/users/services/implementations/UserService.java

```diff
@@ -0,0 +1,177 @@
+package com.imaee.propinq.users.services.implementations;
+
+import com.imaee.propinq.users.controllers.requests.RecoverPasswordRequest;
+import com.imaee.propinq.users.controllers.requests.SendEmailRequest;
+import com.imaee.propinq.users.controllers.requests.SendNewActivationTokenRequest;
+import com.imaee.propinq.users.controllers.requests.SignUpRequest;
+import com.imaee.propinq.users.data.models.Token;
+import com.imaee.propinq.users.data.models.User;
+import com.imaee.propinq.users.data.repositories.IUserRepository;
+import com.imaee.propinq.users.mappers.UserMapper;
+import com.imaee.propinq.users.services.interfaces.IEmailService;
+import com.imaee.propinq.users.services.interfaces.ITokenService;
+import com.imaee.propinq.users.services.interfaces.IUserService;
+import com.imaee.propinq.users.utils.EmailBuilder;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+import org.springframework.web.server.ResponseStatusException;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import java.time.LocalDateTime;
+import java.util.UUID;
+
+@Service
+public class UserService implements IUserService {
+
+    private final IUserRepository userRepository;
+    private final ITokenService tokenService;
+    private final EmailBuilder emailBuilder;
+    private final PasswordEncoder passwordEncoder;
+    private final IEmailService emailService;
+
+    public UserService(IUserRepository userRepository, ITokenService tokenService, EmailBuilder emailBuilder, PasswordEncoder passwordEncoder, IEmailService emailService) {
+        this.userRepository = userRepository;
+        this.tokenService = tokenService;
+        this.emailBuilder = emailBuilder;
+        this.passwordEncoder = passwordEncoder;
+        this.emailService = emailService;
+
+    }
+
+    @Override
+    public void saveUser(SignUpRequest createUserRequest) {
+        ifEmailAlreadyExistsThrowException(createUserRequest.email());
+        ifUsernameAlreadyExistsThrowException(createUserRequest.username());
+        User newUser = createUser(createUserRequest);
+        userRepository.save(newUser);
+        UUID activationTokenId = generateActivationToken(newUser);
+        sendActivationEmail(newUser, activationTokenId);
+    }
+
+    private void ifEmailAlreadyExistsThrowException(String email) {
+        if(userRepository.existsByEmail(email)){
+            throw new IllegalArgumentException("Email already exists.");
+        }
+    }
+
+    private void ifUsernameAlreadyExistsThrowException(String username) {
+        if(userRepository.existsByUsername(username)){
+            throw new IllegalArgumentException("Username already exists.");
+        }
```

Comment:
Lo que te marqué arriba

```suggestion
        if (userRepository.existsByUsername(username))
            throw new IllegalArgumentException("Username already exists.");
```